spinflip/spin_flip_tools.py

```diff
@@ -121,14 +130,76 @@ def __init__(self, gradient_filename, merger_data_loc):
         #need matter part gradients from changing Y_e, n_b
         self.grad_S_R_mat, self.grad_S_L_mat = np.zeros(np.shape(self.grad_S_R_nu)), np.zeros(np.shape(self.grad_S_L_nu))
             
-        #total Sigma Gradients
+        #total Sigma Gradients (spacetime up, spacetime (gradient) low, x,y,z, F, F
         self.grad_S_R, self.grad_S_L = self.grad_S_R_nu + self.grad_S_R_mat, self.grad_S_L_nu + self.grad_S_L_mat
 
+    #total H_L and H_R gradients (ignoring extra terms in inverse p/abs)
+    #(spacetime low, x,y,z, F, F)
+    def grad_H_L(self, theta, phi):
+        basis = Basis(theta,phi)
+        return basis.kappa(self.grad_S_L)
+    def grad_H_R(self, theta, phi):
+        basis = Basis(theta,phi)
+        return basis.kappa(self.grad_S_R)
+   
+        
+    # plot of magnitude of gradient along minimizing resonant direction
+    def minGradients(self, emu_data_loc, p_abs, z = None, phi_resolution = 5):
+        min_gradients = np.zeros((self.gradJ.shape[2], self.gradJ.shape[3], self.gradJ.shape[4])) #x,y,z
+        if z == None:
+            z_range = range(self.gradJ.shape[4])
+        else:
+            z_range = [z]
+        for x in range(self.gradJ.shape[2]):
+            for y in range(self.gradJ.shape[3]):
+                for z in z_range:
+
+                    location = [x + self.limits[0,0],
+                                y + self.limits[1,0],
+                                z + self.limits[2,0]]
+                    emu_filename = emu_data_loc + "i{:03d}".format(location[0])+"_j{:03d}".format(location[1])+"_k{:03d}".format(location[2])+"/allData.h5"
+                    SP = SpinParams(t_sim = self.it, resonance_type = 'simplified', emu_file = emu_filename, merger_data_loc = self.merger_data_loc, gradient_filename = self.gradient_filename, location = location, p_abs=p_abs)
+                    if SP.resonant_theta(phi=0) == None:
+                        min_gradients[x,y,z] = None
+                    else:
+                        gradients = []
+                        for phi in np.linspace(0, 2*np.pi, phi_resolution): 
+                            theta = SP.resonant_theta(phi=phi)
+                            grad_H_L = self.grad_H_L(theta, phi)[:,x,y,z] 
+                            direction = Basis(theta,phi).n_vector
+                            grad_along_direction = np.tensordot(grad_H_L, direction, axes = ([0],[0])) # (F,F)
+                            gradients.append(gm.magnitude(grad_along_direction))
+                        min_gradients[x,y,z] = np.min(gradients)
+        
+        return min_gradients
+    
+    def gradientsPlot(self, emu_data_loc, p_abs, z, vmin = -14, vmax = -9, phi_resolution = 5 , savefig=False, min_gradients = None):
+        if type(min_gradients) == type(None):
```

Comment:
You can also just say `min_gradients == None`

Reply:
Got an error when I did that! numpy doesn't like it when you check truth values of arrays. something like "the truth value of an ndarray is ambiguous"